### PR TITLE
Cranelift: Remove `ABICallee` trait

### PR DESCRIPTION
**Pull request by fitzgen**

### CODE REVIEW

**Review by jameysharp — APPROVED** [review]
Always nice to not only get rid of dynamic dispatch but also heap indirection! This is cool.

I spotted a couple comments that could use a little help, but this looks good to me. I assume Chris should still take a look though.

Naming conventions are hard. `AbiCallee` looks weird to me. But :shrug:

**Comment by bjorn3** [comment]
I thought this trait existed because some architectures may have an ABI sufficiently different from the norm that ABICalleeImpl would not be able to implement it.

File: cranelift/codegen/src/machinst/abi_impl.rs [hunk 1]
```diff
@@ -697,7 +697,7 @@ impl ABISig {
 }
 
 /// ABI object for a function body.
-pub struct ABICalleeImpl<M: ABIMachineSpec> {
+pub struct AbiCallee<M: ABIMachineSpec> {
```
Comment (cfallin) [inline cranelift/codegen/src/machinst/abi_impl.rs:700]: Can we keep the capitalization `ABICallee`? I know there are varying opinions on how camelcase interacts with acronyms; but I find `Abi` for the acronym `ABI` a little unnatural to read. (Sorry for the bikeshed color suggestion!)

**Review by cfallin — APPROVED** [review]
Thanks for tackling this! I'm happy to see things simplified here. LGTM overall, just one slightly bikesheddy comment below.

To address a few points in discussion:

@bjorn3:
> I thought this trait existed because some architectures may have an ABI sufficiently different from the norm that ABICalleeImpl would not be able to implement it.

The trait initially existed because each backend implemented it separately; there was no shared "vanilla ABI" code. Eventually I refactored things to share that (this was in #2128 and #2142 for aarch64 and x64 respectively; and s390x never had its own impl, it has always used `abi_impl`). We kept the trait around because, as you suggest, there was a thought that other machines might need very different implementations.

But in practice, (i) this trait separation adds complexity and hinders refactoring (@fitzgen was hitting some issues recently, motivating this PR), and (ii) we've actually managed to make the shared `abi_impl` code pretty flexible, by adding hooks to the `ABIMachineSpec`. I'm confident that we can tweak things as needed to fit most reasonable architectures -- e.g., the in-progress riscv64 backend uses it with no problem -- and for truly different architectures, we can update the design when we understand the requirements better.

@jameysharp:

> Naming conventions are hard. AbiCallee looks weird to me. But 🤷

Indeed, we spent a bunch of discussion on this two years ago, actually; it was originally some other name that escapes me at the moment (`ABIBody` maybe?) and there was a suggestion to name the traits in terms of the role that they represent. So `ABICallee` is the callee side (receiving arguments, doing prologues and epilogues) and `ABICaller` is the caller side (setting up arguments, emitting a call instruction, noting clobbers as side-effects).

File: cranelift/codegen/src/machinst/abi_impl.rs (hunk 1 above)
Comment (fitzgen) [reply cranelift/codegen/src/machinst/abi_impl.rs:700]: I changed it because it is Rust's naming style to only capitalize the first letter of an acronym in a type (consistent with how you wouldn't add underscores between each letter in the acronym in snake case):

> In <code class="hljs">UpperCamelCase</code>, acronyms and contractions of compound words count as one word: use <code class="hljs">Uuid</code> rather than <code class="hljs">UUID</code>, <code class="hljs">Usize</code> rather than <code class="hljs">USize</code> or <code class="hljs">Stdin</code> rather than <code class="hljs">StdIn</code>. In <code class="hljs">snake_case</code>, acronyms and contractions are lower-cased: <code class="hljs">is_xid_start</code>.

https://rust-lang.github.io/api-guidelines/naming.html

And also because I happen to find it more readable that way as well (no mixing up the "C" in "Callee" as part of the acronym). I plan on doing the rest in follow up PRs that I'm working on.

File: cranelift/codegen/src/machinst/abi_impl.rs [hunk 2]
```diff
@@ -1091,45 +1098,40 @@ impl<M: ABIMachineSpec> ABICallee for ABICalleeImpl<M> {
...
+    /// values. If this function is called multiple times, the maximum of all
+    /// `size` values will be available.
+    pub fn accumulate_outgoing_args_size(&mut self, size: u32) {
         if size > self.outgoing_args_size {
             self.outgoing_args_size = size;
         }
     }
 
-    fn flags(&self) -> &settings::Flags {
-        &self.flags
```
[the remark this answers is not in the document]
Comment (fitzgen) [reply cranelift/codegen/src/machinst/abi_impl.rs:1101]: Yeah I only removed things that were warned as dead code.

File: cranelift/codegen/src/machinst/abi_impl.rs (hunk 1 above)
Comment (jameysharp) [reply cranelift/codegen/src/machinst/abi_impl.rs:700]: When there's a collision between widespread naming conventions and readability, I like to have... both? :sweat_smile:

Is there an alternative name that's meaningful but avoids the "ABI" acronym entirely? Or perhaps, would `CalleeAbi` work as a compromise? Somehow that reads less badly to me.

File: cranelift/codegen/src/machinst/compile.rs [hunk 3]
```diff
@@ -14,7 +14,7 @@ use regalloc2::{self, MachineEnv};
 pub fn compile<B: LowerBackend + TargetIsa>(
     f: &Function,
     b: &B,
-    abi: Box<dyn ABICallee<I = B::MInst>>,
+    abi: AbiCallee<<<B as LowerBackend>::MInst as MachInst>::ABIMachineSpec>,
```
[the remark this answers is not in the document]
Comment (fitzgen) [reply cranelift/codegen/src/machinst/compile.rs:17]: The type checker was complaining it couldn't infer where the `MInst` projection was coming from, not sure why it can't be a little smarter here, but alas.

File: cranelift/codegen/src/machinst/abi_impl.rs (hunk 1 above)
Comment (cfallin) [reply cranelift/codegen/src/machinst/abi_impl.rs:700]: I guess I diverge from the above guidelines then in this particular area :-) This paragraph from the "About" page of that book seems relevant too:

> These are only guidelines, some more firm than others. In some cases they are vague and still in development. Rust crate authors should consider them as a set of important considerations in the development of idiomatic and interoperable Rust libraries, to use as they see fit. These guidelines should not in any way be considered a mandate that crate authors must follow, though they may find that crates that conform well to these guidelines integrate better with the existing crate ecosystem than those that do not.

I find `AbiCallee` looks wrong to me; an ABI is an ABI, not an Abi (or "abi"); I'd really prefer not to take the change. If it's confusing to me and not confusing to you, then there is likely a split among other readers as well. Given that it is currently this way and there is disagreement, could we keep with the current convention for this PR and then discuss separately? (I'd the same concern if the other names were also modified in this way.)

Comment (cfallin) [reply cranelift/codegen/src/machinst/abi_impl.rs:700]: I saw @jameysharp 's reply after writing mine -- but this brings to mind another option: just `Caller` and `Callee`? (or `CallerHandler` and `CalleeHandler`?)

Comment (jameysharp) [reply cranelift/codegen/src/machinst/abi_impl.rs:700]: Ooh, I like `abi::Callee`.

File: cranelift/codegen/src/isa/aarch64/abi.rs [hunk 4]
```diff
@@ -65,7 +65,7 @@ fn saved_reg_stack_size(
 
 /// AArch64-specific ABI behavior. This struct just serves as an implementation
 /// point for the trait; it is never actually instantiated.
-pub(crate) struct AArch64MachineDeps;
+pub struct AArch64MachineDeps;
```
Comment (bjorn3) [inline cranelift/codegen/src/isa/aarch64/abi.rs:68]: Why this change?

Comment (fitzgen) [reply cranelift/codegen/src/isa/aarch64/abi.rs:68]: Because it is public now that there is the `MachInst::ABIMachineSpec` associated type.
